### PR TITLE
[QUIC] Fixed disabled receives if buffer gets drained before RECEIVE returns

### LINKED ISSUE #72196: [Quic] System.Net.Quic.Tests.MsQuicTests.WriteTests timeout in rolling build
[Build](https://runfo.azurewebsites.net/view/build/?number=1880188), [Log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-86f871399ad4442b86/System.Net.Quic.Functional.Tests/2/console.ab8d2f0f.log?%3Fhelixlogtype%3Dresult)

```
 System.Net.Quic.Functional.Tests: [Long Running Test] 'System.Net.Quic.Tests.MsQuicTests.WriteTests', Elapsed: 00:02:11
    System.Net.Quic.Tests.MsQuicTests.WriteTests(writes: [[1000004, 502, 15003, 1, 502, ...], [15003, 1, 15003, 1, 1000004, ...], [1, 1, 1, 1, 502, ...], [1, 1000004, 1000004, 1, 1000004, ...], [1, 502, 15003, 1000004, 502, ...]]) [FAIL]
      System.TimeoutException : The operation has timed out.
      Stack Trace:
        /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs(55,0): at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks, Int32 millisecondsTimeout)
        /_/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs(281,0): at System.Net.Quic.Tests.QuicTestBase.RunClientServer(Func`2 clientFunction, Func`2 serverFunction, Int32 iterations, Int32 millisecondsTimeout, QuicListenerOptions listenerOptions)
        /_/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs(281,0): at System.Net.Quic.Tests.QuicTestBase.RunClientServer(Func`2 clientFunction, Func`2 serverFunction, Int32 iterations, Int32 millisecondsTimeout, QuicListenerOptions listenerOptions)
        /_/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs(281,0): at System.Net.Quic.Tests.QuicTestBase.RunClientServer(Func`2 clientFunction, Func`2 serverFunction, Int32 iterations, Int32 millisecondsTimeout, QuicListenerOptions listenerOptions)
        /_/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs(281,0): at System.Net.Quic.Tests.QuicTestBase.RunClientServer(Func`2 clientFunction, Func`2 serverFunction, Int32 iterations, Int32 millisecondsTimeout, QuicListenerOptions listenerOptions)
        /_/src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs(593,0): at System.Net.Quic.Tests.MsQuicTests.WriteTests(Int32[][] writes)
        --- End of stack trace from previous location ---
```

Runfo Tracking Issue: [system.net.quic.tests.msquictests.writetests](https://runfo.azurewebsites.net/tracking/issue/361)
|Build|Definition|Kind|Run Name|Console|Core Dump|Test Results|Run Client|
|---|---|---|---|---|---|---|---|
|[1882195](https://dev.azure.com/dnceng/public/_build/results?buildId=1882195)|[runtime](https://dnceng.visualstudio.com/public/_build?definitionId=686)|[PR 72235](https://github.com/dotnet/runtime/pull/72235)|net7.0-windows-Release-x86-CoreCLR_release-Windows.10.Amd64.Server2022.ES.Open|[console.log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72235-merge-62029cf31de1407ba7/System.Net.Quic.Functional.Tests/2/console.7aa0a3a8.log?%3F%253Fhelixlogtype%253Dresult)|||[runclient.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72235-merge-62029cf31de1407ba7/System.Net.Quic.Functional.Tests/b27b087e-e429-45ed-935b-aff45697d059.log?%3F%253Fhelixlogtype%253Dresult)|
|[1882195](https://dev.azure.com/dnceng/public/_build/results?buildId=1882195)|[runtime](https://dnceng.visualstudio.com/public/_build?definitionId=686)|[PR 72235](https://github.com/dotnet/runtime/pull/72235)|net7.0-windows-Release-x86-CoreCLR_release-Windows.10.Amd64.Server2022.ES.Open|[console.log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72235-merge-62029cf31de1407ba7/System.Net.Quic.Functional.Tests/2/console.7aa0a3a8.log?%3F%253Fhelixlogtype%253Dresult)|||[runclient.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72235-merge-62029cf31de1407ba7/System.Net.Quic.Functional.Tests/b27b087e-e429-45ed-935b-aff45697d059.log?%3F%253Fhelixlogtype%253Dresult)|
|[1881865](https://dev.azure.com/dnceng/public/_build/results?buildId=1881865)|[runtime](https://dnceng.visualstudio.com/public/_build?definitionId=686)|[PR 72129](https://github.com/dotnet/runtime/pull/72129)|net7.0-windows-Debug-x64-CoreCLR_release-Windows.11.Amd64.Client.Open|[console.log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72129-merge-3da47510349842ba85/System.Net.Quic.Functional.Tests/2/console.f9c115d5.log?%3F%253Fhelixlogtype%253Dresult)|||[runclient.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72129-merge-3da47510349842ba85/System.Net.Quic.Functional.Tests/758b98f5-15d1-431f-b024-89a139918f99.log?%3F%253Fhelixlogtype%253Dresult)|
|[1881865](https://dev.azure.com/dnceng/public/_build/results?buildId=1881865)|[runtime](https://dnceng.visualstudio.com/public/_build?definitionId=686)|[PR 72129](https://github.com/dotnet/runtime/pull/72129)|net7.0-windows-Debug-x64-CoreCLR_release-Windows.11.Amd64.Client.Open|[console.log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72129-merge-3da47510349842ba85/System.Net.Quic.Functional.Tests/2/console.f9c115d5.log?%3F%253Fhelixlogtype%253Dresult)|||[runclient.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72129-merge-3da47510349842ba85/System.Net.Quic.Functional.Tests/758b98f5-15d1-431f-b024-89a139918f99.log?%3F%253Fhelixlogtype%253Dresult)|
|[1881761](https://dev.azure.com/dnceng/public/_build/results?buildId=1881761)|[runtime](https://dnceng.visualstudio.com/public/_build?definitionId=686)|Rolling|net7.0-windows-Release-x64-CoreCLR_release-Windows.10.Amd64.Server2022.ES.Open|[console.log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-70-preview71d9aab21c3324bcba8/System.Net.Quic.Functional.Tests/2/console.68b9c1ed.log?%3F%253Fhelixlogtype%253Dresult)|||[runclient.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-70-preview71d9aab21c3324bcba8/System.Net.Quic.Functional.Tests/889b132c-e011-465d-859e-6f0e7f676380.log?%3F%253Fhelixlogtype%253Dresult)|
|[1881761](https://dev.azure.com/dnceng/public/_build/results?buildId=1881761)|[runtime](https://dnceng.visualstudio.com/public/_build?definitionId=686)|Rolling|net7.0-windows-Release-x64-CoreCLR_release-Windows.10.Amd64.Server2022.ES.Open|[console.log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-70-preview71d9aab21c3324bcba8/System.Net.Quic.Functional.Tests/2/console.68b9c1ed.log?%3F%253Fhelixlogtype%253Dresult)|||[runclient.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-release-70-preview71d9aab21c3324bcba8/System.Net.Quic.Functional.Tests/889b132c-e011-465d-859e-6f0e7f676380.log?%3F%253Fhelixlogtype%253Dresult)|
|[1880188](https://dev.azure.com/dnceng/public/_build/results?buildId=1880188)|[runtime](https://dnceng.visualstudio.com/public/_build?definitionId=686)|Rolling|net7.0-windows-Release-x64-CoreCLR_release-Windows.11.Amd64.Client.Open|[console.log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-86f871399ad4442b86/System.Net.Quic.Functional.Tests/2/console.ab8d2f0f.log?%3F%253Fhelixlogtype%253Dresult)|||[runclient.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-86f871399ad4442b86/System.Net.Quic.Functional.Tests/f647e099-a7d4-47dd-b85b-581ed81605d9.log?%3F%253Fhelixlogtype%253Dresult)|
|[1880188](https://dev.azure.com/dnceng/public/_build/results?buildId=1880188)|[runtime](https://dnceng.visualstudio.com/public/_build?definitionId=686)|Rolling|net7.0-windows-Release-x64-CoreCLR_release-Windows.11.Amd64.Client.Open|[console.log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-86f871399ad4442b86/System.Net.Quic.Functional.Tests/2/console.ab8d2f0f.log?%3F%253Fhelixlogtype%253Dresult)|||[runclient.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-86f871399ad4442b86/System.Net.Quic.Functional.Tests/f647e099-a7d4-47dd-b85b-581ed81605d9.log?%3F%253Fhelixlogtype%253Dresult)|
|[1878908](https://dev.azure.com/dnceng/public/_build/results?buildId=1878908)|[runtime](https://dnceng.visualstudio.com/public/_build?definitionId=686)|[PR 72116](https://github.com/dotnet/runtime/pull/72116)|net7.0-windows-Debug-x64-CoreCLR_release-Windows.10.Amd64.Server2022.ES.Open|[console.log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72116-merge-b49c9a7018834aff93/System.Net.Quic.Functional.Tests/2/console.04baf1c1.log?%3F%253Fhelixlogtype%253Dresult)|||[runclient.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72116-merge-b49c9a7018834aff93/System.Net.Quic.Functional.Tests/aea45343-8dac-4528-aeb8-9442c13fba26.log?%3F%253Fhelixlogtype%253Dresult)|
|[1878908](https://dev.azure.com/dnceng/public/_build/results?buildId=1878908)|[runtime](https://dnceng.visualstudio.com/public/_build?definitionId=686)|[PR 72116](https://github.com/dotnet/runtime/pull/72116)|net7.0-windows-Debug-x64-CoreCLR_release-Windows.10.Amd64.Server2022.ES.Open|[console.log](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72116-merge-b49c9a7018834aff93/System.Net.Quic.Functional.Tests/2/console.04baf1c1.log?%3F%253Fhelixlogtype%253Dresult)|||[runclient.py](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-72116-merge-b49c9a7018834aff93/System.Net.Quic.Functional.Tests/aea45343-8dac-4528-aeb8-9442c13fba26.log?%3F%253Fhelixlogtype%253Dresult)|

Build Result Summary
|Day Hit Count|Week Hit Count|Month Hit Count|
|---|---|---|
|0|5|5|

### LINKED ISSUE #72422: Test failure System.Net.Quic.Tests.MsQuicTests.ByteMixingOrNativeAVE_MinimalFailingTest
Run: [runtime-libraries-coreclr outerloop 20220718.5](https://dev.azure.com/dnceng/public/_build/results?buildId=1887157&view=ms.vss-test-web.build-test-results-tab&runId=49281830&paneView=debug&resultId=104544)

Failed test:
```
net7.0-windows-Release-x64-CoreCLR_release-Windows.Amd64.Server2022.Open

- System.Net.Quic.Tests.MsQuicTests.ByteMixingOrNativeAVE_MinimalFailingTest

```
**Error message:**
```
System.TimeoutException : The operation has timed out.


Stack trace
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks, Int32 millisecondsTimeout) in /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 55
   at System.Net.Quic.Tests.QuicTestBase.RunClientServer(Func`2 clientFunction, Func`2 serverFunction, Int32 iterations, Int32 millisecondsTimeout, QuicListenerOptions listenerOptions) in /_/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs:line 299
   at System.Net.Quic.Tests.QuicTestBase.RunClientServer(Func`2 clientFunction, Func`2 serverFunction, Int32 iterations, Int32 millisecondsTimeout, QuicListenerOptions listenerOptions) in /_/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs:line 299
   at System.Net.Quic.Tests.QuicTestBase.RunClientServer(Func`2 clientFunction, Func`2 serverFunction, Int32 iterations, Int32 millisecondsTimeout, QuicListenerOptions listenerOptions) in /_/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs:line 299
   at System.Net.Quic.Tests.QuicTestBase.RunClientServer(Func`2 clientFunction, Func`2 serverFunction, Int32 iterations, Int32 millisecondsTimeout, QuicListenerOptions listenerOptions) in /_/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicTestBase.cs:line 299
   at System.Net.Quic.Tests.MsQuicTests.<ByteMixingOrNativeAVE_MinimalFailingTest>g__RunTest|23_0(Byte[] data) in /_/src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:line 838
   at System.Threading.Tasks.TaskTimeoutExtensions.GetRealException(Task task) in /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 128
--- End of stack trace from previous location ---
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks) in /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 92
   at System.Threading.Tasks.TaskTimeoutExtensions.WhenAllOrAnyFailed(Task[] tasks, Int32 millisecondsTimeout) in /_/src/libraries/Common/tests/System/Threading/Tasks/TaskTimeoutExtensions.cs:line 55
   at System.Net.Quic.Tests.MsQuicTests.ByteMixingOrNativeAVE_MinimalFailingTest() in /_/src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:line 873
--- End of stack trace from previous location ---
```

### PR DESCRIPTION
**Pull request by ManickaP**
Root cause: we're reading in a tight loop, so when RECEIVE event arrived, we'd copy some data (based on buffer capacity) and return SUCCESS with partial receive. But the reading loop thread was faster and it would drain the buffer before the event handler returned to MsQuic. Now we have `ReadAsync` waiting on the `_receiveTcs` (waiting on another RECEIVE event) and MsQuic waiting on `StreamReceiveSetEnabled` since we did partial data receive.

The fix: check if the receive buffer has a capacity (i.e.: someone read from it in the mean time) and if so immediately re-enable receives with `QUIC_STATUS_CONTINUE` 

Interestingly, this manifested only on Windows, so I ran the offending test in a tight loop for ~10 minutes without any failures (normally it'd manifest in less than 1 minute).

Fixes #72196
Fixes #72422

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs [hunk 1]
```diff
@@ -509,7 +509,7 @@ private unsafe int HandleEventReceive(ref RECEIVE data)
         _receiveTcs.TrySetResult();
 
         data.TotalBufferLength = totalCopied;
-        return QUIC_STATUS_SUCCESS;
+        return _receiveBuffers.HasCapacity() ? QUIC_STATUS_CONTINUE : QUIC_STATUS_SUCCESS;
```
[the remark this answers is not in the document]
Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:512]: I see. The (initial) problem happens when this line on Read thread
https://github.com/dotnet/runtime/blob/c6f538e5c1eb4c2472d043bf835936aa4f372ff2/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs#L310

executes earlier than this line on MsQuic thread
https://github.com/dotnet/runtime/blob/83ec9470869a0806d8cfc35ba40826197c2f5c28/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs#L506

meaning we are able to catch that situation any time after Volatile.Write, including on return.

I have one more question:
Is it ok when we both return CONTINUE and also schedule EnableReceive afterwards? What do you think about including the same CompareExchange here as well:
```suggestion
        return (_receiveBuffers.HasCapacity() && Interlocked.CompareExchange(ref _receivedNeedsEnable, 0, 1) == 1) ? QUIC_STATUS_CONTINUE : QUIC_STATUS_SUCCESS;
```

[the remark this answers is not in the document]
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:512]: Hmmm, let me check if that works without breaking anything.
That would prevent queuing `StreamReceiveSetEnabled` if we've already re-enabled receives via CONTINUE status --> -1 mindless P/Invoke.

[the remark this answers is not in the document]
Comment (ManickaP) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs:512]: So far, looks good, I'll update the PR.

**Review by CarnaViire — APPROVED** [review]
Thanks!

**Comment by ManickaP** [comment]
Failures look like https://github.com/dotnet/runtime/issues/72429, i.e.: `-1073740791` exit code on remote executors.

**Pull request merged**